datalad/cmd.py

```diff
@@ -6,66 +6,152 @@
+
+        *args and **kwargs:
+           see Runner.run() and Runner.call() for available arguments.
 
-        Provides improved logging for debugging purposes and raises
-        RuntimeError exception in case of non-0 return
+        Raises
+        ------
+        ValueError if cmd is neither a string nor a callable.
         """
+
+        if isinstance(cmd, basestring):
+            self.run(cmd, *args, **kwargs)
+        elif callable(cmd):
+            self.call(cmd, *args, **kwargs)
+        else:
+            raise ValueError("Argument 'command' is neither a string nor a callable.")
+
+    def run(self, cmd, log_stdout=False, log_stderr=True):
+        """Runs the command `cmd` using shell.
+
+        In case of dry-mode `cmd` is just added to `commands` and it is executed otherwise.
+        Allows for seperatly logging stdout and stderr  or streaming it to system's stdout
+        or stderr respectively.
+
+        Parameters
+        ----------
+        cmd : str
+          String defining the command call.
+
+        log_stdout: bool
+            If True, stdout is logged. Goes to sys.stdout otherwise.
+
+        log_stderr: bool
+            If True, stderr is logged. Goes to sys.stderr otherwise.
+
+        Returns
+        -------
+        Status code as returned by the called command or `None` in case of dry-mode.
+
+        Raises
+        ------
+        RunTimeError
+           if command's exitcode wasn't 0 or None
+        """
+
+        if log_stdout:
+            outputstream = subprocess.PIPE
+        else:
+            outputstream = sys.stdout
+
+        if log_stderr:
+            errstream = subprocess.PIPE
+        else:
+            errstream = sys.stderr
+
         self.log("Running: %s" % (cmd,))
+
         if not self.dry:
-            # status, output = commands.getstatusoutput(cmd)
-            # doing manually for improved debugging
-            pipe = os.popen('{ ' + cmd + '; } 2>&1', 'r')
-            output = ''
-            for line in iter(pipe.readline, ''):
-                self.log("| " + line.rstrip('\n'), level=5)
-                output += line
-            if output[-1:] == '\n': output = output[:-1]
-            status = pipe.close() or 0
-
-            if not status in [0, None]:
-                msg = "Failed to run %r. Exit code=%d output=%s" \
-                      % (cmd, status, output)
+
+            proc = subprocess.Popen(cmd, stdout=outputstream, stderr=errstream, shell=True)
+            # shell=True allows for piping, multiple commands, etc., but that implies to not use shlex.split()
+            # and is considered to be a security hazard. So be careful with input.
+            # Alternatively we would have to parse `cmd` and create multiple subprocesses.
+
+            while proc.poll() is None:
+                if log_stdout:
+                    line = proc.stdout.readline()
+                    if line != '':
```

Comment:
and bloody wget ... how it dare to log its madness to stderr for no good faulty reason!
so now it results for us an "error" stderr being logged whenever there is no error... damn it

btw -- now, if I run test_main.py tests I see stderr logged only with empty line

```
2015-03-09 22:55:24,551 [DEBUG  ] Running: cd /home/yoh/.tmp/tmpEXX2rT && git annex addurl  -c annex.alwayscommit=false --file "files/1.tar.gz"  "http://localhost:8004/.tmp-page2annex-nXQ1jA/1.tar.gz" (cmd.py:172)
2015-03-09 22:55:24,625 [DEBUG  ] HTTP: "HEAD /.tmp-page2annex-nXQ1jA/1.tar.gz HTTP/1.1" 200 - (utils.py:165)
addurl files/1.tar.gz (downloading http://localhost:8004/.tmp-page2annex-nXQ1jA/1.tar.gz ...) 
2015-03-09 22:55:24,647 [DEBUG  ] HTTP: "GET /.tmp-page2annex-nXQ1jA/1.tar.gz HTTP/1.1" 200 - (utils.py:165)
2015-03-09 22:55:24,648 [ERROR  ] stderr|  (cmd.py:172)
ok
(recording state in git...)
2015-03-09 22:55:24,657 [ERROR  ] stderr|      0K  
```
but if I am replicating this via cmdline  (```chmod +w -R /tmp/123 && rm -rf /tmp/123 || : ; mkdir -p /tmp/123; cd /tmp/123; git init; git annex init; git annex addurl -c annex.alwayscommit=false --file "1.tar.gz" "http://onerussian.com/tmp/1.tar.gz" >|/tmp/1-stdout 2>|/tmp/1-stderr | cat ; du -scb /tmp/1-stderr;```) -- or even just directly via Runner.run() 
```
$> DATALAD_LOGLEVEL=10 python -c "from datalad.cmd import Runner; Runner().run('chmod +w -R /tmp/123 && rm -rf /tmp/123 || : ; mkdir -p /tmp/123; cd /tmp/123; git init; git annex init; git annex addurl -c annex.alwayscommit=false --file "1.tar.gz" "http://onerussian.com/tmp/1.tar.gz"', log_stdout=True)"         
2015-03-09 22:59:47,426 [DEBUG  ] Running: chmod +w -R /tmp/123 && rm -rf /tmp/123 || : ; mkdir -p /tmp/123; cd /tmp/123; git init; git annex init; git annex addurl -c annex.alwayscommit=false --file 1.tar.gz http://onerussian.com/tmp/1.tar.gz (cmd.py:172)
2015-03-09 22:59:47,432 [DEBUG  ] stdout| Initialized empty Git repository in /tmp/123/.git/ (cmd.py:172)
2015-03-09 22:59:47,639 [ERROR  ] stderr|  (cmd.py:172)
2015-03-09 22:59:47,640 [DEBUG  ] stdout| init  ok (cmd.py:172)
2015-03-09 22:59:47,667 [ERROR  ] stderr|      0K .......... .......... .......... .......... ..........  1% 1.78M 2s (cmd.py:172)
2015-03-09 22:59:47,667 [DEBUG  ] stdout| (recording state in git...) (cmd.py:172)
2015-03-09 22:59:47,687 [ERROR  ] stderr|     50K .......... .......... .......... .......... ..........  2% 2.44M 2s (cmd.py:172)
2015-03-09 22:59:48,887 [DEBUG  ] stdout| addurl 1.tar.gz (downloading http://onerussian.com/tmp/1.tar.gz ...)  (cmd.py:172)
2015-03-09 22:59:48,887 [ERROR  ] stderr|    100K .......... .......... .......... .......... ..........  4% 2.08M 2s (cmd.py:172)
2015-03-09 22:59:48,888 [DEBUG  ] stdout| ok (cmd.py:172)
2015-03-09 22:59:48,888 [ERROR  ] stderr|    150K .......... .......... .......... .......... ..........  5% 2.65M 1s (cmd.py:172)
2015-03-09 22:59:48,888 [DEBUG  ] stdout| (recording state in git...) (cmd.py:172)
2015-03-09 22:59:48,888 [ERROR  ] stderr|    200K .......... .......... .......... .......... ..........  7% 3.01M 1s (cmd.py:172)
2015-03-09 22:59:48,891 [ERROR  ] stderr|    250K .......... .......... .......... .......... ..........  8% 2.34M 1s (cmd.py:172)
```
stderr is getting filled up with progress indication for wget... do you know why it is so so, ben?